testctrl/svc/operations_server.go

```diff
@@ -16,29 +16,259 @@ package svc
+	var time *timestamppb.Timestamp
+	if e == nil {
+		return
+	}
+	time, err = ptypes.TimestampProto(e.Time)
+	if err != nil {
+		err = fmt.Errorf(
+			"cannot convert event timestamp to proto: %s",
+			err,
+		)
+		return
+	}
+	eventpb = &svcpb.Event{
+		Subject: e.SubjectName,
+		Kind:    e.Kind.Proto(),
+		Time:    time,
+	}
+	return
+}
+
+// newOperationMetadataProto constructs an operation metadata proto from
+// an Event object. The operation metadata proto is of type TestSessionMetadata,
+// marshalled to Any. This function returns an error if the metadata proto
+// cannot be constructred or cannot be marshalled to Any.
+func newOperationMetadataProto(e *types.Event) (metadatapb *anypb.Any, err error) {
+	var eventpb *svcpb.Event
+	eventpb, err = newEventProto(e)
+	if err != nil {
+		return
+	}
+	metadatapb, err = ptypes.MarshalAny(&svcpb.TestSessionMetadata{
+		LatestEvent:    eventpb,
+		ServiceVersion: Version,
+	})
+	if err != nil {
+		err = fmt.Errorf("could not format test metadata: %s", err)
+	}
+	return
+}
+
+// newOperationResponseProto constructs an operation response proto from an
+// Event object and a Session object. The Session object is only needed for the
+// creation time. The operation response proto is of type TestSessionResult,
+// marshalled to Any. This function returns an error if the response proto
+// cannot be constructed or cannot be marshalled to Any.
+func newOperationResponseProto(e *types.Event, s *types.Session) (responsepb *anypb.Any, err error) {
+	if (e == nil) || (!isOperationDone(e)) || (s == nil) {
+		err = fmt.Errorf("cannot calculate operation response: failed precondition check")
+		return
+	}
+	responsepb, err = ptypes.MarshalAny(&svcpb.TestSessionResult{
+		DriverLogs:  e.DriverLogs,
+		TimeElapsed: ptypes.DurationProto(s.CreateTime.Sub(e.Time)),
+	})
+	if err != nil {
+		err = fmt.Errorf(
+			"could not format test session result into operation response: %v",
+			err,
+		)
+	}
+	return
+}
+
+// newOperationErrorProto constructs an operation error proto from an Event
+// object. The error response proto is of type Status. This function returns
+// an error if the error proto cannot be constructed.
+func newOperationErrorProto(e *types.Event) (status *statuspb.Status, err error) {
+	if (e == nil) || (!isOperationDone(e)) || isOperationSuccess(e) {
+		err = fmt.Errorf("cannot calculate operation error: failed precondition check")
+		return
+	}
+	var statusCode int32
+	switch e.Kind {
+	case types.InternalErrorEvent:
+		statusCode = int32(codepb.Code_INTERNAL)
+	default:
+		statusCode = int32(codepb.Code_UNKNOWN)
+	}
+	status = &statuspb.Status{
+		Code:    statusCode,
+		Message: e.Description,
+	}
+	return
+}
+
+// newOperation constructs an operation proto from an Event object and a Session
+// object. This function returns en error if the operation proto cannot be
+// constructed.
+func newOperation(e *types.Event, s *types.Session) (o *lrpb.Operation, err error) {
+	if s == nil {
```

Comment:
would we want to check that event is not nil as well?

Reply:
I see... at first glance, I did not realize `isOperationDone` handles nil and checks for completion.

Reply:
Made changes to clarify the code and distinguish finished/done.